Fix: Accessibility issue with links listing blocks. #(64051)

im3dabasia opened the pull request:
<!-- Thanks for contributing to Gutenberg! Please follow the Gutenberg Contributing Guidelines:
https://github.com/WordPress/gutenberg/blob/trunk/CONTRIBUTING.md -->

## What?
Fixes #64051.

## Why?
This PR addresses accessibility by ensuring screen readers recognize these blocks as navigational elements, as per issue #64051.

## How?

Added `<nav> `wrappers around the lists in `core/page-list` and `core/tag-cloud` blocks. For `core/categories`, the `<nav>` is conditionally added when displayed as a list (not as a dropdown), improving accessibility.

## Screenshots or screencast 

| Before | After |
|--------|-------|
| ![Before](https://github.com/user-attachments/assets/8897aa49-accd-4d6d-9d9f-7f65f43e3e83) | ![After](https://github.com/user-attachments/assets/a7a68510-3928-443e-b571-6d980eef5737) |

im3dabasia committed:
Fix: Wrap link lists in <nav> for accessibility in listing blocks (#64051)

carolinan commented:
Hi @im3dabasia I do not believe that this is the correct solution for this problem. I also believe that the problem and use case needs to be described in more detail before a solution can be proposed. I recommend allowing some more time to discuss the issue first.

The `<nav>` element is intended to be used for navigations, and all lists of links are not necessarily navigations.

carolinan commented on packages/block-library/src/categories/edit.js:
```diff
@@ -257,9 +257,11 @@ export default function CategoriesEdit( {
+					renderCategoryDropdown()
+				) : (
+					<nav>{ renderCategoryList() }</nav>
+				) ) }
```
By placing the `<nav>` inside `<TagName>` the resulting markup is a `<nav>` inside a `<ul>`, which is not valid HTML.

Instead, the `TagName` constant should be updated and the block props should be placed on the new outer element, the `nav.`
This is how the `TagName` currently sets the tag:
```
	const TagName =
		!! categories?.length && ! displayAsDropdown && ! isResolving
			? 'ul'
			: 'div';

```

carolinan commented on packages/block-library/src/categories/index.php:
```diff
@@ -66,7 +66,7 @@ function render_block_core_categories( $attributes, $content, $block ) {
 		$args['show_option_none'] = $taxonomy->labels->no_terms;
 
-		$wrapper_markup = '<ul %1$s>%2$s</ul>';
+		$wrapper_markup = '<nav><ul %1$s>%2$s</ul></nav>';
```
The block props need to be on the outer element.

carolinan commented on packages/block-library/src/page-list/edit.js:
```diff
@@ -109,7 +109,11 @@ function BlockContent( {
+			<nav>
+				<ul { ...innerBlocksProps }></ul>
+			</nav>
+		);
```
I would recommend excluding this block from the PR since the page list can be used in the navigation block. Adding the nav here invalidates the HTML and may break the navigation block in unexpected ways.

carolinan commented on packages/block-library/src/tag-cloud/index.php:
```diff
@@ -41,7 +41,7 @@ function render_block_core_tag_cloud( $attributes ) {
 
 	return sprintf(
-		'<p %1$s>%2$s</p>',
+		'<nav %1$s>%2$s</nav>',
```
This works well in my test, but the nav needs a unique label. It also needs to take into the account that more than one of copy of the tag cloud block can be used on the same page.

im3dabasia replied:
Hey @carolinan ,

I have addressed the changes you suggested earlier. Could you please guide me on the changes I need to make based on this comment? I am unsure.

im3dabasia commented:
Thank you for the feedback, @carolinan! I'll hold off on any additional changes until there's further discussion on the issue's scope and proposed solution.

carolinan commented:
> Thank you for the feedback, @carolinan! I'll hold off on any additional changes until there's further discussion on the issue's scope and proposed solution.

My concern is the page list block, if that can be separated I think work on the other two blocks can continue.

im3dabasia commented:
> > Thank you for the feedback, @carolinan! I'll hold off on any additional changes until there's further discussion on the issue's scope and proposed solution.
> 
> My concern is the page list block, if that can be separated I think work on the other two blocks can continue.

Thank you for the reply @carolinan.

I will start working on the other 2 blocks but I need some guidance. In the category list block. The suggestion was to change the constant tagName to 'nav' instead of 'ul', Here I had a doubt since the function renderCategoryList() returns a array of `<li>`( renderCategoryList calls renderCategoryListItem() internally). If I change the TagName to 'nav' then the generated markup would be 

```
<nav>
<li> </li>
<li> </li>
<li> </li>
</nav>
```

This would result in wrong markup. I am confused here on how to approach this. Any guidance on this would be much appreciated

carolinan commented:
@im3dabasia I think you can change the `<nav>` element that you added around `renderCategoryList()` to a `<ul>`, and then change `<ul>` in the  TagName constant to a nav?

im3dabasia committed:
fix: Remove pagelist block nav changes

im3dabasia committed:
fix: Add props to outer element in categories list and fixed the html markup

carolinan commented:
I found that the tag cloud has a `format` parameter where developers can choose between an unordered list and a flat list of items. But both formats can be inside a `<nav>` without any validation problems. https://developer.wordpress.org/reference/functions/wp_tag_cloud/

There aren't many examples of adding aria labels that are editable by the user. The navigation block uses one, and there is a suggestion to add one to the group block.

1. `ariaLabel` is a block support, it needs to be set to true in block.json.
2. After updating the block support, the documentation also needs to be updated. `npm run docs:build`.
3. The option needs to be a text input field in the Advanced section of the InspectorControls.
4. The option needs a suitable default value for when the user does not customize the label.
5. The option needs a suitable label. The label for the option in the navigation block is "Menu name" , so perhaps "Tag Cloud name"?



The categories are different because it already has a label text that is used when the dropdown option is selected.
This label text could be re-used as the aria label on the nav.

im3dabasia commented:
Hey @carolinan 

I am new to contributing and just want to verify if I need to do the following:

a) Enable ariaLabel support in block.json.
b) Add a text input field for the aria-label in the Advanced section of InspectorControls, ensuring it has appropriate default values and labels.
c) Update the documentation after the changes by running npm run docs:build.

Please let me know if I’m on the right track or if there’s anything I’m missing.

carolinan commented:
Yes.
But also be prepared that there may be more discussions and reviews since this is not a common pattern in the existing blocks.

im3dabasia committed:
feat: Add aria label textcontrol for tag cloud block in inspector controls

im3dabasia commented:
Hi @carolinan,

I have implemented the requested changes and added support for the aria-label, allowing the user to customize it. By default, the name is set to "Tag Cloud", but the user has the option to modify it in the Advanced section of the Inspector Controls.


Attaching a video reference for the same:

https://github.com/user-attachments/assets/0b7c882b-f751-4632-8030-feec79673e49

Please take a look at the changes, and I would appreciate any feedback you may have.

Thank you!

carolinan commented on packages/block-library/src/tag-cloud/index.php:
```diff
@@ -17,6 +17,7 @@
 function render_block_core_tag_cloud( $attributes ) {
 	$smallest_font_size = $attributes['smallestFontSize'];
 	$unit               = ( preg_match( '/^[0-9.]+(?P<unit>[a-z%]+)$/i', $smallest_font_size, $m ) ? $m['unit'] : 'pt' );
+	$aria_label         = isset( $attributes['ariaLabel'] ) ? $attributes['ariaLabel'] : 'Tag Cloud';
```
The default text must be translatable, placed inside a translation function.

carolinan commented:
Please solve the merge conflict first :)

im3dabasia committed:
fix: Default tag cloud name in translation function

im3dabasia commented:
@carolinan I have solved the merge conflicts. Please check again. 🙇

carolinan commented:
When I try to checkout the branch it gets rejected: ` ! [rejected]   `
I don't know how to fi x that. Maybe there are still conflicts?

carolinan commented:
There is a valid error reported regarding the Unit tests because the test fixtures have not been updated with the new block attribute.
Try `npm run fixtures:regenerate`

carolinan commented on packages/block-library/src/tag-cloud/block.json:
```diff
@@ -28,6 +28,10 @@
+		},
+		"ariaLabel": {
+			"type": "string",
+			"default": "Tag Cloud"
```
I am a bit rusty after being away for several weeks but I don't think that having a default text here will work because It needs to be translatable?

t-hamano replied:
With the merge of #69002, this custom attribute is no longer necessary. Can you rebase this PR and then remove this attribute?

im3dabasia replied:
Hey @t-hamano, thank you for #69002. I appreciate your efforts in helping with this PR.

After observing what you did in [#68611](https://github.com/WordPress/gutenberg/pull/68611/files), I aligned the Tag Cloud block in a similar manner. I removed the ariaLabel attribute from the attributes section and used supports instead.

The Tag Cloud block is rendered on the server side using the ServerSideRender component, and I have relevant PHP code in the index.php file of the Tag Cloud block to handle the aria-label attribute in the HTML output.

The aria-label is added on the frontend, but I'm not quite sure why this error is appearing. I'd appreciate it if you could help with this.

After making these changes, I'm now seeing this error in my dev console. 

https://github.com/user-attachments/assets/8e9c566a-996c-48fb-85b1-8b4752ae25b6



```json
{
    "code": "rest_invalid_param",
    "message": "Invalid parameter(s): attributes",
    "data": {
        "status": 400,
        "params": {
            "attributes": "ariaLabel is not a valid property of Object."
        },
        "details": {
            "attributes": {
                "code": "rest_additional_properties_forbidden",
                "message": "ariaLabel is not a valid property of Object.",
                "data": null
            }
        }
    }
}
```

Whenever you have a moment, could you take a look?

im3dabasia committed:
fix: Remove ariaLabel attribute

t-hamano commented:
[This comment](https://github.com/WordPress/gutenberg/pull/66866#discussion_r1943363050) made me realize that the `aria-label` support is not complete. We are currently discussing how to address this in #68764.

---

About this PR: My biggest concern is backwards compatibility.

This PR changes the structure of the output HTML, which may have a significant impact on themes that already apply styles to these blocks via stylesheets.

#### If we replace the block wrapper element itself with a nav element

The HTML will change as follows:

```html
<!-- From -->
<p class="wp-block-tag-cloud">
	<a href="">Tag</a>
</p>

<!-- To -->
<nav class="wp-block-tag-cloud">
	<a href="">Tag</a>
</nav>
```

Selectors like `p.wp-block-cloud` will no longer work. 

#### If we wrap the block wrapper element in a `nav` element

The HTML will change as follows:

```html
<!-- From -->
<p class="wp-block-tag-cloud">
	<a href="">Tag</a>
</p>

<!-- To -->
<nav>
	<p class="wp-block-tag-cloud">
		<a href="">Tag</a>
	</p>
</nav>
```

Selectors like `.entry-content > .wp-block-cloud` will no longer work. This may also affect margin support, layout, etc.

With this in mind, I think we need to have more discussion before moving forward, taking into account the impact on backward compatibility.

carolinan commented:
Its not great but the accessibility is more important.

carolinan commented:
I did a quick search on wpdirectory.net, which does not replace doing more research:
`.wp-block-tag-cloud` is in 1559 themes.
`p.wp-block-tag-cloud` is in 281 themes and 1 plugin.

t-hamano commented on packages/block-library/src/tag-cloud/edit.js:
```diff
@@ -123,125 +125,143 @@ function TagCloudEdit( { attributes, setAttributes } ) {
+				<TextControl
+					__next40pxDefaultSize
+					__nextHasNoMarginBottom
+					label={ __( 'Tag Cloud Name' ) }
```
```suggestion
					label={ __( 'Tag cloud name' ) }
```

t-hamano commented on packages/block-library/src/tag-cloud/index.php:
```diff
@@ -40,9 +41,12 @@ function render_block_core_tag_cloud( $attributes ) {
+
 	return sprintf(
-		'<p %1$s>%2$s</p>',
+		'<nav %1$s %2$s>%3$s</nav>',
 		$wrapper_attributes,
+		$aria_label_attribute,
```
It should be possible to write it as follows:

```php
$wrapper_attributes = get_block_wrapper_attributes( array( 'aria-label' => $aria_label ) );

return sprintf(
	'<nav %1$s>%2$s</nav>',
	$wrapper_attributes,
	$tag_cloud
);
```

t-hamano reviewed (commented):
If we proceed with this PR, I recommend that you thoroughly investigate the impact of this change, along with accessibility feedback. For example:

- Is the global styles applied correctly?
- Are there any visual changes in the default theme?
- Are block support, especially spacing support, applied correctly?

Additionally, we should consider the impact of selectors like the following if a theme uses them:

- `.wp-block-categories-list > li`
- `.wp-block-categories > li`
- `.wp-block-tag-cloud > a`

im3dabasia committed:
Merge branch 'trunk' into fix/64051-accessibility-links-listing-blocks

im3dabasia committed:
fix: Improve code quality in ariaLabel php code

t-hamano commented:
I want to reaffirm whether moving forward with this PR approach is truly important from an accessibility perspective.

This PR will result in the following changes to the HTML:

#### Tag Cloud Block

```html
<!-- From -->
<p class="wp-block-tag-cloud">
  <a href="" class="tag-cloud-link" aria-label="Tag A (1 item)">Tag A</a>
  <a href="" class="tag-cloud-link" aria-label="Tag B (1 item)">Tag B</a>
  <a href="" class="tag-cloud-link" aria-label="Tag C (1 item)">Tag C</a>
</p>

<!-- To -->
<nav class="wp-block-tag-cloud" aria-label="Tag Cloud">
  <a href="" class="tag-cloud-link" aria-label="Tag A (1 item)">Tag A</a>
  <a href="" class="tag-cloud-link" aria-label="Tag B (1 item)">Tag B</a>
  <a href="" class="tag-cloud-link" aria-label="Tag C (1 item)">Tag C</a>
</nav>
```

#### Categories Block (as a list)

```html
<!-- From -->
<ul class="wp-block-categories-list wp-block-categories">
  <li class="cat-item cat-item-1"><a href="">Category A</a></li>
  <li class="cat-item cat-item-2"><a href="">Category B</a></li>
  <li class="cat-item cat-item-3"><a href="">Category C</a></li>
</ul>

<!-- To -->
<nav class="wp-block-categories-list wp-block-categories">
  <ul>
    <li class="cat-item cat-item-1"><a href="">Category A</a></li>
    <li class="cat-item cat-item-2"><a href="">Category B</a></li>
    <li class="cat-item cat-item-3"><a href="">Category C</a></li>
  </ul>
</nav>
```

My concerns are:

- Do these blocks need to be wrapped in a `nav` element?
- Are `nav` elements without `aria-label` acceptable?
- If `nav` elements without `aria-label` are not acceptable, do we need to expose the `aria-label` editing UI?
- Blocks may be duplicated and have the same `aria-label`.

cc @afercia @joedolson @alexstine

joedolson commented:
If a block is duplicated and has the same label, that's OK; a block with the same content is semantically the same block, so a duplicate label is accurate. 

I would say that yes, they should be wrapped in `nav` elements. Like @afercia says, having feature compatibility with the core widgets is the base line, and those are wrapped in `nav` elements. 

However, I note that while block themes *mostly* declare HTML5 by default, they are omitting the navigation widgets from that declaration; is that an oversight?

See: r52439 & https://github.com/WordPress/wordpress-develop/blob/b5e0c16c04edcebdf5e81c39796a9d2c22cfd6f7/src/wp-includes/theme.php#L4394

carolinan commented:
> If a block is duplicated and has the same label, that's OK; a block with the same content is semantically the same block, so a duplicate label is accurate.

Aha, yes that makes sense, thank you,

carolinan commented:
> However, I note that while block themes _mostly_ declare HTML5 by default, they are omitting the navigation widgets from that declaration; is that an oversight?
> 
> See: [r52439](https://core.trac.wordpress.org/changeset/52439) & https://github.com/WordPress/wordpress-develop/blob/b5e0c16c04edcebdf5e81c39796a9d2c22cfd6f7/src/wp-includes/theme.php#L4394

I am not sure what a navigation widget is.

afercia commented:
> I am not sure what a navigation widget is.

`navigation-widgets` is one of the extra arguments for `add_theme_support( 'html5, $args )`. When supported, many classic widgets like the navigation menu widget, archives, categories, pages, recent posts, recent comments etc. wrap their html output within a labeled `<nav>` element. 

As @joedolson pointed out, it appears block themes do not support the full `html5 support' extra arguments as `navigation-widgets` is missing from [the extra arga array here](https://github.com/WordPress/wordpress-develop/blob/ac057e6c692e7e88b1758eb8dc6dee2312bb5c7b/src/wp-includes/theme.php#L4389-L4394).

carolinan commented:
But where is it documented? it is not listed here under the html5 heading. https://developer.wordpress.org/reference/functions/add_theme_support/

carolinan commented:
Yeah I would assume it was unintentionally missed because it is not clearly documented.

afercia commented:
> Yeah I would assume it was unintentionally missed because it is not clearly documented.

I guess so. I see the REST API doesn't support it as well? This new support was intended to be fully 'opt-in' but I'm not sure why it hasn't been documented.

t-hamano commented:
Thanks everyone for your feedback!

> I would say that at the very least the related blocks should provide feature parity with what was implemented in Core, although the Core implementation is behind opting-in for `html5` support for `navigation-widgets`. Reference:

Does this mean that if a theme doesn't have `html5` theme support, we shouldn't use nav elements?

If so, I'm wondering if this argument applies to other blocks as well. For example:

- Navigation and Query Pagination blocks always use `nav` elements:
  - https://github.com/WordPress/gutenberg/blob/c196e9a183c645167fb7f520cbd19c3026290912/packages/block-library/src/navigation/index.php#L710
  -  https://github.com/WordPress/gutenberg/blob/c196e9a183c645167fb7f520cbd19c3026290912/packages/block-library/src/query-pagination/index.php#L32
- A PR has been submitted to make the nav element selectable in the group block: https://github.com/WordPress/gutenberg/pull/68611

carolinan commented:
> Does this mean that if a theme doesn't have html5 theme support, we shouldn't use nav elements?

No, the blocks should use modern HTML that is supported by the browsers that WordPress supports.

carolinan commented:
The HTML theme support exists to make it easier for theme developers to use modern HTML, while keeping the features backwards compatible with older themes that has styles that expects a certain markup. As seen here: https://core.trac.wordpress.org/ticket/61183

afercia commented:
> Does this mean that if a theme doesn't have html5 theme support, we shouldn't use nav elements?

Ideally, blocks themes (and thus the blocks they use) should output modern html5 markup and use best practices by default. as @carolinan pointed out, the `html5` support was introduced many years ago to allow themes to progressively, granularly, introduce new html5 features. I think blocks should support all these html5 features by default.

t-hamano commented:
Sorry for the late reply.

I thought about how we could move this PR forward. I think it may be good to focus on **just adding a wrapper `nav` element with the hard-coded `aria-label` text**.

That is, remove the `aria-label` support from the Tag Cloud block. There is no block that exposes the aria-label editing UI, so it would need to be discussed more deeply. Additionally, I'm not sure if any users would want to change the aria-abel text, as both blocks have an obvious purpose.

I think the key here is probably to see what visual changes occur in the default themes.

Let me know what you think!
